[JSC] Optimize `Array#lastIndexOf` in DFG/FTL

sosukesuzuki opened the pull request:
#### 4a99a165c178cf6ec0a8ce5bc1ab3bea1ba859f2
<pre>
[JSC] Optimize `Array#lastIndexOf` in DFG/FTL
<a href="https://bugs.webkit.org/show_bug.cgi?id=300170">https://bugs.webkit.org/show_bug.cgi?id=300170</a>

Reviewed by NOBODY (OOPS!).

This reverts commit <a href="https://commits.webkit.org/302763@main">302763@main</a> to reland <a href="https://commits.webkit.org/302157@main">302157@main</a>.

When `fromIndex &gt;= array.length`, the code uses `emitPopulateSliceIndex`
which clamps large values to length. However, `lastIndexOf` should start
searching from `length - 1`, not length. This caused the loop to access
`array[length]`, which is out of bounds.

So this new patch changes to add additional clamping logic specifically
for `lastIndexOf` to ensure that when `fromIndex &gt;= length`, it gets
clamped to `length - 1` instead of `length`.

-----
(original commit message)

This patch changes to optimize `Array#lastIndexOf` in DFG/FTL.

                                                  TipOfTree                  Patched

array-prototype-lastIndexOf-string-16           0.7393+-0.0963            0.6353+-0.0238          might be 1.1637x faster
array-prototype-lastIndexOf-double-from-contiguous
                                                0.7476+-0.3292            0.6312+-0.0984          might be 1.1845x faster
array-prototype-lastIndexOf-int32               3.8388+-0.8260     ^      2.3827+-0.3433        ^ definitely 1.6111x faster
array-prototype-lastIndexOf-bigint              9.9618+-0.7504     ^      8.3997+-0.4838        ^ definitely 1.1860x faster
array-prototype-lastIndexOf-contiguous          5.4971+-1.0560     ^      2.1412+-0.1266        ^ definitely 2.5673x faster
array-prototype-lastIndexOf-string-const        3.7607+-0.1203     ^      2.9016+-0.1071        ^ definitely 1.2961x faster
array-prototype-lastIndexOf-untyped-int32
                                               80.1997+-0.6415     ^     42.8832+-0.8174        ^ definitely 1.8702x faster
array-prototype-lastIndexOf-int32-from-contiguous
                                                0.6262+-0.1345            0.6172+-0.0809          might be 1.0146x faster
array-prototype-lastIndexOf-double              2.3190+-0.0583     ^      1.3878+-0.0684        ^ definitely 1.6711x faster
array-prototype-lastIndexOf-string              0.7423+-0.1096            0.6556+-0.1410          might be 1.1323x faster

* JSTests/stress/array-prototype-lastIndeOf-large-fromIndex.js: Added.
(main.v125):
(main.v85):
(main.v58):
(main):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
(JSC::DFG::FixupPhase::fixupArrayIndexOfOrArrayIncludes):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasArrayMode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::arrayLastIndexOfString):
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileArrayIndexOfOrArrayIncludes):
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
* Source/JavaScriptCore/runtime/Intrinsic.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a99a165c178cf6ec0a8ce5bc1ab3bea1ba859f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130416 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2687 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41370 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137834 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82008 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2692 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2579 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99367 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67220 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133363 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1969 "Found 1 new test failure: http/tests/site-isolation/history/add-iframes-and-navigate-mainframe.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116788 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80065 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1894 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34918 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81093 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122419 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110446 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35423 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140311 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/128869 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2477 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2271 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107879 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2521 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113130 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107782 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1930 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31573 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55436 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2547 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65935 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/161884 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2365 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40370 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2568 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2473 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -9658,10 +9659,21 @@ void SpeculativeJIT::compileArrayIndexOfOrArrayIncludes(Node* node)
-    else
-        move(TrustedImm32(0), indexGPR);
+        if (isArrayLastIndexOf) {
+            auto indexInRange = branch32(LessThan, indexGPR, lengthGPR);
+            move(lengthGPR, indexGPR);
+            sub32(TrustedImm32(1), indexGPR);
+            indexInRange.link(this);
+        }
```
change from original patch 1

sosukesuzuki commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -8305,23 +8306,30 @@ IGNORE_CLANG_WARNINGS_END
-            startIndex = m_out.int32Zero;
-
-        bool isArrayIncludes = m_node->op() == ArrayIncludes;
+            if (isArrayLastIndexOf)
+                startIndex = m_out.select(m_out.greaterThanOrEqual(startIndex, length), m_out.sub(length, m_out.int32One), startIndex);
```
change from original patch 2

sosukesuzuki reviewed (commented):
The only changes from the original patch are the two points I commented on.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -9658,10 +9659,21 @@ void SpeculativeJIT::compileArrayIndexOfOrArrayIncludes(Node* node)
+        if (isArrayLastIndexOf) {
+            auto indexInRange = branch32(LessThan, indexGPR, lengthGPR);
+            move(lengthGPR, indexGPR);
+            sub32(TrustedImm32(1), indexGPR);
```
Use, `sub32(lengthGPR, TrustedImm32(1), indexGPR)`

sosukesuzuki replied:
Fixed!

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -9658,10 +9659,21 @@ void SpeculativeJIT::compileArrayIndexOfOrArrayIncludes(Node* node)
+    } else {
+        if (isArrayLastIndexOf) {
+            move(lengthGPR, indexGPR);
+            sub32(TrustedImm32(1), indexGPR);
```
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -9822,7 +9851,7 @@ void SpeculativeJIT::compileArrayIndexOfOrArrayIncludes(Node* node)
 
         if (isCopyOnWriteArrayWithContiguous()) {
-            operation = operationCopyOnWriteArrayIndexOfString;
+            indexOfOperation = operationCopyOnWriteArrayIndexOfString;
```
Is it right? This does not consider about lastIndex.

sosukesuzuki replied:
This is simply a rename of `operation` to `indexOfOperation`. This is to make explicit that it's not an operation for `includes` and `lastIndexOf`.

Constellation replied:
I mean, we should not take this path when it is lastIndexOf.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -9926,7 +9962,7 @@ void SpeculativeJIT::compileArrayIndexOfOrArrayIncludes(Node* node)
         } else {
             addSlowPathGenerator(slowPathCall(
-                slowCase, this, operation,
+                slowCase, this, isArrayLastIndexOf ? operationArrayLastIndexOfString : indexOfOperation,
```
This looks not correct. I think lastIndexOf case needs multiple operations?

sosukesuzuki replied:
Do you mean we should have a operation like `operationCopyOnWriteArrayIndexOfString` for `lastIndexOf` as well as `indexOf`?

I had not intended to support `lastIndexOf` optimizations for CoW Arrays in this patch.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -8444,7 +8455,7 @@ IGNORE_CLANG_WARNINGS_END
 
             if (isCopyOnWriteArrayWithContiguous()) {
-                operation = operationCopyOnWriteArrayIndexOfString;
+                indexOfOperation = operationCopyOnWriteArrayIndexOfString;
```
Ditto.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -8526,7 +8540,7 @@ IGNORE_CLANG_WARNINGS_END
 
             m_out.appendTo(slowCase, continuation);
-            ValueFromBlock slowCaseResult = isArrayIncludes ? m_out.anchor(vmCall(Int32, operationArrayIncludesString, weakPointer(globalObject), storage, searchElement, startIndex)) : m_out.anchor(vmCall(Int64, operation, weakPointer(globalObject), storage, searchElement, startIndex));
+            ValueFromBlock slowCaseResult = isArrayIncludes ? m_out.anchor(vmCall(Int32, operationArrayIncludesString, weakPointer(globalObject), storage, searchElement, startIndex)) : m_out.anchor(vmCall(Int64, isArrayLastIndexOf ? operationArrayLastIndexOfString : indexOfOperation, weakPointer(globalObject), storage, searchElement, startIndex));
```
Ditto.

sosukesuzuki committed:
[JSC] Optimize `Array#lastIndexOf` in DFG/FTL
https://bugs.webkit.org/show_bug.cgi?id=300170

Reviewed by NOBODY (OOPS!).

This reverts commit 302763@main to reland 302157@main.

When `fromIndex >= array.length`, the code uses `emitPopulateSliceIndex`
which clamps large values to length. However, `lastIndexOf` should start
searching from `length - 1`, not length. This caused the loop to access
`array[length]`, which is out of bounds.

So this new patch changes to add additional clamping logic specifically
for `lastIndexOf` to ensure that when `fromIndex >= length`, it gets
clamped to `length - 1` instead of `length`.

-----
(original commit message)

This patch changes to optimize `Array#lastIndexOf` in DFG/FTL.

                                                  TipOfTree                  Patched

array-prototype-lastIndexOf-string-16           0.7393+-0.0963            0.6353+-0.0238          might be 1.1637x faster
array-prototype-lastIndexOf-double-from-contiguous
                                                0.7476+-0.3292            0.6312+-0.0984          might be 1.1845x faster
array-prototype-lastIndexOf-int32               3.8388+-0.8260     ^      2.3827+-0.3433        ^ definitely 1.6111x faster
array-prototype-lastIndexOf-bigint              9.9618+-0.7504     ^      8.3997+-0.4838        ^ definitely 1.1860x faster
array-prototype-lastIndexOf-contiguous          5.4971+-1.0560     ^      2.1412+-0.1266        ^ definitely 2.5673x faster
array-prototype-lastIndexOf-string-const        3.7607+-0.1203     ^      2.9016+-0.1071        ^ definitely 1.2961x faster
array-prototype-lastIndexOf-untyped-int32
                                               80.1997+-0.6415     ^     42.8832+-0.8174        ^ definitely 1.8702x faster
array-prototype-lastIndexOf-int32-from-contiguous
                                                0.6262+-0.1345            0.6172+-0.0809          might be 1.0146x faster
array-prototype-lastIndexOf-double              2.3190+-0.0583     ^      1.3878+-0.0684        ^ definitely 1.6711x faster
array-prototype-lastIndexOf-string              0.7423+-0.1096            0.6556+-0.1410          might be 1.1323x faster

* JSTests/stress/array-prototype-lastIndeOf-large-fromIndex.js: Added.
(main.v125):
(main.v85):
(main.v58):
(main):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
(JSC::DFG::FixupPhase::fixupArrayIndexOfOrArrayIncludes):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasArrayMode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::arrayLastIndexOfString):
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileArrayIndexOfOrArrayIncludes):
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
* Source/JavaScriptCore/runtime/Intrinsic.h:
